gapic/src/main/com/google/gapid/models/Devices.java

```diff
@@ -295,4 +315,89 @@ public boolean isStadia() {
       return device.getConfiguration().getOS().getKind() == Device.OSKind.Stadia;
     }
   }
+
+  /*
+   *  Class that handles the gapis communication for device validation and the load/store
+   *  of the validation cache.
+   */
+  private static class DeviceValidationInfo {
+    private final Client client;
+    private DeviceValidation.Builder deviceValidation;
+
+    public DeviceValidationInfo(Client client, Settings settings) {
+      this.client = client;
+      deviceValidation = settings.deviceValidation();
+    }
+
+    private static DeviceValidation.ValidationEntry buildValidationEntry(DeviceCaptureInfo device) {
+      return DeviceValidation.ValidationEntry.newBuilder()
+          .setDevice(DeviceValidation.Device.newBuilder()
+              .setSerial(device.device.getSerial())
+              .setOs(device.device.getConfiguration().getOS())
+              .setVersion(device.device.getConfiguration().getDrivers().getVulkan().getVersion()))
+          .setResult(DeviceValidation.Result.newBuilder()
+              .setPassed(true))
+          .build();
+    }
+
+    public DeviceValidationResult getValidationStatus(DeviceCaptureInfo device) {
+      if (skipDeviceValidation.get()) {
+        return DeviceValidationResult.getSkippedResult();
+      }
+      DeviceValidation.ValidationEntry entry  = buildValidationEntry(device);
+      if (deviceValidation.getValidationEntriesList().contains(entry)) {
+        return DeviceValidationResult.getPassedResult();
+      }
+      return DeviceValidationResult.getFailedResult();
+    }
+
+    public ListenableFuture<DeviceValidationResult> doValidation(DeviceCaptureInfo device) {
+      if (device == null) {
+        return immediateFuture(DeviceValidationResult.getPassedResult());
+      }
+      DeviceValidation.ValidationEntry currentEntry = buildValidationEntry(device);
+      if (deviceValidation.getValidationEntriesList().contains(currentEntry)) {
+        return immediateFuture(DeviceValidationResult.getPassedResult());
+      }
+
+      return MoreFutures.transform(client.validateDevice(device.path), e -> {
+        updateValidationStatus(currentEntry, e);
+        return new DeviceValidationResult(e.getError(), !e.hasError(), false);
+      });
+    }
+
+    protected void updateValidationStatus(DeviceValidation.ValidationEntry entry, Service.ValidateDeviceResponse response) {
```

Comment:
this needs to happen on the UI thread.

Reply:
This is just for updating the cache that a new entry is going to be added. It will reach this step only if the device wasn't in the cache already. The Devices model will handle the proto response message while constructing a new ValidationResult object and pass that to the UI.